docs/source/hallway.rst

```diff
@@ -0,0 +1,363 @@
+state index.
+
+In our example above, this method will yield
+
+::
+
+   * [new(1)] if cur_pos == 0
+   * [new(0), new(2)] if cur_pos == 1
+   * [] if cur_pos == 2
+
+Note that the number of children (actions) is variable, and that the
+terminal state returns an empty list of next children.
+
+``_make_observation``
+~~~~~~~~~~~~~~~~~~~~~
+
+To decide which child to transition to, the RL agent will need to call a
+policy model with that vertex’s observation. To this end, we implement
+``_make_observation`` which, for the hallway example, returns:
+
+.. code:: python
+
+   def _make_observation(self) -> Dict[str, np.ndarray]:
+       """Makes an observation of this HallwayState vertex.
+       Returns:
+           Dict[str, np.ndarray]: dictionary containing the current position
+           index under the key 'cur_pos'.
+       """
+       return {
+           "cur_pos": np.array([self.cur_pos], dtype=int),
+       }
+
+Note that the returned observation must exactly match the specification
+in the vertex’s ``observation_space``.
+
+HallwayModel
+------------
+
+(See ``graphenv.examples.hallway.hallway_model`` for the full
+implementation).
+
+The ``Model`` class implements the policy model used by the RL algorithm
+and, as such, needs to be implemented to take vertex observation data as
+input, and to output an action value and action weight for each
+observation. In practice, this amounts to implementing a keras model in
+the ``__init__``, and storing it in the ``base_model`` attribute of the
+model class.
+
+.. code:: python
+
+   class HallwayModel(GraphModel):
+       """An example GraphModel implementation for the HallwayEnv and HallwayState
+       Graph.
+       Attributes:
+           base_model : The Keras model used to evaluate vertex observations.
+       """
+
+       def __init__(
+           self,
+           *args,
+           hidden_dim: int = 1,
+           **kwargs,
+       ):
+           """Initializs this HallwayModel.
+           Uses a dense fully connected Keras network.
+           Args:
+               hidden_dim (int, optional): The number of hidden layers to use. 
+                   Defaults to 1.
+           """
+           super().__init__(*args, **kwargs)
+
+           cur_pos = layers.Input(shape=(1,), name="cur_pos", dtype=tf.float32)
+
+           hidden_layer = layers.Dense(hidden_dim, name="hidden_layer")
+           action_value_output = layers.Dense(
+               1, name="action_value_output", bias_initializer="ones"
+           )
+           action_weight_output = layers.Dense(
+               1, name="action_weight_output", bias_initializer="ones"
+           )
+
+           out = hidden_layer(cur_pos)
+           action_values = action_value_output(out)
+           action_weights = action_weight_output(out)
+
+           self.base_model = tf.keras.Model(
+               [cur_pos], [action_values, action_weights])
+
+HallwayEnv
+----------
+
+TODO: HERE
```

Comment:
the only other concept I think we need to hit is the `max_num_actions` idea, where we're handling the padding / masking, but everything in `Vertex` is written from the point of view of a *single* instance.

Reply:
yeah exactly, will be sure to mention